modules/apm/NAMING.md

```diff
@@ -1,82 +1,115 @@
 
-Attribute names should follow the same rules. In particular, these rules apply to attributes too:
-* elements and separators
-* hierarchy/namespaces
-* units
-* pluralization (when an attribute represents a measurement)
 
-For **pluralization**, when an attribute represents an entity, the attribute name should be singular (e.g.` es.security.realm_type`, not` es.security.realms_type` or `es.security.realm_types`), unless it represents a collection (e.g.` es.rest.request_headers`)
+## Metric attributes
 
+**WARNING** Do not use **high cardinality** attributes / dimensions. This might result in the APM Java agent dropping events.
 
-### List of previously registered metric names
-You can inspect all previously registered metrics names with
-`./gradlew run -Dtests.es.logger.org.elasticsearch.telemetry.apm=debug`
-This should help you find out the already registered group that your meteric
-might fit
+It is not always straight forward to decide if something should part of the metric name or an attribute (dimension) of that metric.  As a rule of thumb:
+- any aggregation across any dimensions of a metric should be meaningful, and
+- meaningful aggregations should be possible without having to aggregate over different metrics.
+
+Naming conventions for attributes are very similar to metric names.
+Always use `es_` as the root segment to easily discover ES attributes.
+Follow with a module name, team or area of code, e.g. `snapshot, repositories, indices, threadpool` using existing terminology (whether singular and plural).
+
+```
+es(_<segment>)+
+```
+
+- Each **segment** should be lower-case ascii and digits only;
+    - max 30 characters per segment are permitted,
+    - at least two segments are required and
+    - there's a limit of 8 segments per metric name.
+- The segment **separator** is the underscore (`_`).
+- The name may not exceed 255 characters.
+
+Attributes that represent an entity should be named in singular.
+If the attribute value represents a collection, it should be named in plural.
+
+### Migration of existing, invalid attributes
+
+Unfortunately, validation of attribute names was only introduced retrospectively.
+Existing attributes, that do not comply with naming conventions, are tracked in a skip list to not fail validation.
+However, their usage will be logged when running with assertions enabled.
+
+Please migrate usage of such attributes following these steps:
+1. Check if the attribute is used on dashboards or alerts (elasticsearch-o11y-resources). If not, rename the attribute following above conventions; otherwise continue.
+2. Duplicate the attribute using a name following above conventions.
+3. Once there's sufficient history, update alerts and dashboards to use the new attribute.
+4. Remove the old attribute from the skip list and delete all usages.
+
+## Inspect registered metrics
+
+To inspect currently registered metrics, run ES using:
+```
+./gradlew run -Dtests.es.logger.org.elasticsearch.telemetry.apm=debug
```

Comment:
I believe this will emit the metric names as they are registered. My suggestion above is to dump these in an ordered/hierarchical fashion to make visual inspection possible.